File: Linked List/SinglyLinkedList.js
// Singly Linked List and Some common patterns to solve problems

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  // Push Front (as Head)
  pushFront(val) {
    const newNode = new Node(val);
    newNode.next = this.head;
    this.head = newNode;
  }

  // Push back (as Tail)
  pushBack(val) {
    const newNode = new Node(val);

    if (this.head == null) {
      this.head = newNode;
      return;
    }

    let curr = this.head;
    while (curr.next) {
      curr = curr.next;
    }

    curr.next = newNode;
  }

  showList() {
    let curr = this.head;
    const elements = [];

    while (curr) {
      elements.push(curr.val);
      curr = curr.next;
    }

    console.log(elements);
  }

  reverseList() {
    let prev = null;
    let curr = this.head;

    while (curr) {
      const next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }
    this.head = prev;
  }
}

const list = new SinglyLinkedList();
list.pushFront(30);
list.pushBack(40);
list.pushBack(50);
list.pushFront(20);
list.pushBack(60);
list.pushFront(10);
list.showList();
list.reverseList();
list.showList();

File: Linked List/SinglyLinkedList.go
// Singly Linked List and Some common patterns to solve problems

package main

import "fmt"

type Node struct {
	val  int
	next *Node
}

func Constructor() *Node {
	return nil
}

// Push Front (as Head)
func pushFront(head *Node, val int) *Node {
	newNode := &Node{val: val}
	newNode.next = head
	return newNode
}

// Push back (as Tail)
func pushBack(head *Node, val int) *Node {
	newNode := &Node{val: val}

	if head == nil {
		return newNode
	}

	curr := head
	for curr.next != nil {
		curr = curr.next
	}

	curr.next = newNode
	return head
}

func showList(head *Node) {
	curr := head
	elements := []int{}

	for curr != nil {
		elements = append(elements, curr.val)
		curr = curr.next
	}

	fmt.Println(elements)
}

func reverseList(head *Node) *Node {
	var prev *Node
	curr := head

	for curr != nil {
		next := curr.next
		curr.next = prev
		prev = curr
		curr = next
	}
	return prev
}

func main() {
	var head *Node

	head = pushFront(head, 20)
	head = pushBack(head, 30)
	head = pushBack(head, 40)
	head = pushFront(head, 10)

	showList(head)

	head = reverseList(head)
	showList(head)
}
